### PR TITLE
Typed Bucket Replication Support

### PR DESCRIPTION
**Pull request by bowrocker**
#### Overview

Riak 2.0 brings with it typed buckets -- the ability to create a type and associate a bucket with that type:

https://github.com/basho/riak/issues/362

In order to successfully replicate a typed object from one cluster to another, the type definition must exist and be equal on both clusters. MDC replication must handle the possibility that the type of a given object exists on the replication source cluster, but not on the sink cluster (repl 2 terminology).

At this time, there is no facility to automatically create type across clusters -- for example, create a type that does not exist on the sink cluster -- to facilitate replication happening seemlessly. This could be an avenue to explore.

Additionally, replication must be extended to properly handle typed buckets. This work is assumed to come along with support for typed-bucket replication. "Default" type buckets (legacy buckets in Riak) continue to be automatically replicated as before.

This document discusses 2 possible options for implementing this in replication:

https://gist.github.com/bowrocker/7f0d9d6879493f1ac0e9

This PR implements the second option.
##### Type Checking on the Sink

This option allows typed buckets to be replicated to the sink cluster without any checking on the source. During the do_repl_put on the sink, the bucket is checked. If it has a type, that type is checked for existence, and if it does not exist, the object is dropped, and an error message is printed, and preferably an alarm of some sort is sent.

This option assumes that types should exist on both source and sink clusters, and that the non-existance of a type is an operational error that the user or CSE should take care of. Once the type is created, object of that type will replicate correctly.
###### Pros
- This is the simplest solution. Very few moving parts, very low impact on operations and performance.
###### Cons
- Bandwidth, RTQ space, and processing is wasted when an object whose type does not exist on the sink cluster is replicated.
- The assumption that all types should exist on both sides.
#### Configuration

No new user configuration is introduced. If the sink is not configured with the type of an object being replicated, an error message is printed. It is the responsibility of the user to provision this type on the sink. 
##### Mixed version replication

A new replication protocol version is introduced for bucket type support: {2,1}. The following is supported for each version:

| Source | Sink | Result |
| --- | :-: | --: |
| {2,1} | {2,1} | "default" and user defined typed buckets replicated |
| {2,1} | {2,0} or less | only "default" buckets replicated |
#### Dependencies

This PR depends upon code in the following riak_kv branch to be merged, or the conditional post-commit hook present will not work:

https://github.com/basho/riak_repl/tree/jdb-conditional-postcommit

### CODE REVIEW

File: src/riak_repl_util.erl [hunk 1]
```diff
@@ -108,44 +108,62 @@ get_partitions(Ring) ->
...
+    Key = riak_object:key(Object),
     case repl_helper_recv(Object) of
         ok ->
-            ReqId = erlang:phash2({self(), os:timestamp()}),
-            B = riak_object:bucket(Object),
-            K = riak_object:key(Object),
-            Opts = [asis, disable_hooks, {update_last_modified, false}],
+            case Bucket of
+               {T,B} ->
+                   case riak_core_bucket_type:status(T) of
```
Comment (jrwest) [inline src/riak_repl_util.erl]: I may be missing some other area that does the additional checking but isn't it also necessary to check the bucket types are equal\* here? Only checking for existence doesn't prevent us from replicating bucket type `X` with `consistent=true` on the source to the sink who has bucket type `X` set to `consistent=false, datatype=map`. 

\* where equal means some subset of "important" properties are the same on both sides of the cluster?

Comment (bowrocker) [inline src/riak_repl_util.erl]: Agreed. I would think the ones you mention: strongly consistent objects, and CRDTs, at the least.

Is there an example of this sort of checking anywhere? I am guessing what needs to happen is the properties of the incoming bucket type need to be compared with the properties of the type on the sink...

Comment (jrwest) [inline src/riak_repl_util.erl]: The only checking I can think of is `riak_kv_bucket:validate` but thats not totally the same. There are certainly some other properties we care about as well (n_val? allow_mult?). I guess we could implement it w/o knowing the set of properties we need to check upfront (just do all?) and before merge we can get the proper list. Thoughts?

File: src/riak_repl2_rt.erl [hunk 2]
```diff
@@ -143,15 +143,16 @@ postcommit(RObj) ->
...
         Objects0 when is_list(Objects0) ->
             Objects = Objects0 ++ [RObj],
+            Meta = check_for_typed_bucket(RObj),
             BinObjs = riak_repl_util:to_wire(w1, Objects),
             %% try the proxy first, avoids race conditions with unregister()
             %% during shutdown
             case whereis(riak_repl2_rtq_proxy) of
                 undefined ->
-                    riak_repl2_rtq:push(length(Objects), BinObjs);
+                    riak_repl2_rtq:push(length(Objects), BinObjs, Meta);
```
Comment (bookshelfdave) [inline src/riak_repl2_rt.erl:152]: I admit I haven't looked, but can the call to riak_repl2_rtq:push/3 break backwards compat w/ repl 1.3 + 1.4?

File: src/riak_repl2_rt.erl [hunk 3]
```diff
@@ -217,3 +216,13 @@ do_ring_trans(F, A) ->
         ER ->
             ER
     end.
+
+check_for_typed_bucket(Obj) ->
+    M = orddict:new(),
+    case riak_object:bucket(Obj) of
+        {_T, _B } ->
+            lager:info("found typed bucket, setting meta data"),
```
Comment (bookshelfdave) [inline src/riak_repl2_rt.erl]: lager:debug

File: src/riak_repl2_rtsink_conn.erl [hunk 4]
```diff
@@ -54,7 +54,7 @@
 
 %% Register with service manager
 register_service() ->
-    ProtoPrefs = {realtime,[{2,0}, {1,4}, {1,1}, {1,0}]},
+    ProtoPrefs = {realtime,[{2,1}, {2,0}, {1,4}, {1,1}, {1,0}]},
```
Comment (bookshelfdave) [inline src/riak_repl2_rtsink_conn.erl:57]: general comment, we should document what each version does :-(
not required for a PR +1 though
I see you have it in the PR writeup, maybe we should keep track of these elsewhere?

File: src/riak_repl_util.erl [hunk 5]
```diff
@@ -108,44 +108,62 @@ get_partitions(Ring) ->
...
+                       undefined ->
+                           lager:error("Bucket type ~p not defined on sink!", [T]);
```
Comment (bookshelfdave) [inline src/riak_repl_util.erl]: probably don't need the !'s, customers will shit their pants regardless

File: src/riak_repl_util.erl [hunk 6]
```diff
@@ -108,44 +108,62 @@ get_partitions(Ring) ->
...
+                       created ->
+                           lager:error("Bucket type ~p exists on sink, but is not active!", [T]);
+                       ready ->
+                           lager:error("Bucket type ~p exists on sink, but is not active!", [T]);
+                       active ->
+                           lager:debug("Bucket type ~p of object ~p found on sink and is active.", [B, T]),
+                           do_put(B, Object)
```
Comment (bookshelfdave) [inline src/riak_repl_util.erl]: anything you need to match against for `do_put`?

File: src/riak_repl_util.erl [hunk 7]
```diff
@@ -108,44 +108,62 @@ get_partitions(Ring) ->
             riak_core_ring:all_owners(riak_core_ring:upgrade(Ring))]).
 
 do_repl_put(Object) ->
```
Comment (bookshelfdave) [inline src/riak_repl_util.erl:110]: does `do_repl_put` need to return an explicit value?

**Comment by bowrocker** [comment]
Updated for comments and rebased. New PR:

https://github.com/basho/riak_repl/pull/486

**Pull request closed without merging**